### DISCUSSION TITLE
SessionNotOnOrAfter

### DISCUSSION THREAD

**Discussion by larsivi** [node-saml/node-saml#264] [Q&A]
Hi!

I am integrating node-saml in a system that is using accounts-js for authentication/session management (not all users will use SAML/SSO).

During testing I note that logout/session validity is a somewhat varying concept, e.g. SAML via Google Workspace does not provide a logout endpoint. Doing session-checking by redirection feels a bit disruptive (especially as the application is a SPA) and is potentially slow.

With for example Keycloak, I receive a SessionNotOnOrAfter, but are there a way to get that back out from node-saml? I do understand that I can tease this out from the response I give to node-saml, but this seems a useful thing to be able to get in the validate callback or similar.

**Comment by srd90** [comment]
>With for example Keycloak, I receive a SessionNotOnOrAfter, but are there a way to get that back out from node-saml? I do understand that I can tease this out from the response I give to node-saml, but this seems a useful thing to be able to get in the validate callback or similar.

If you are referring to this:
```plain
Response/Assertion/AuthnStatement#SessionNotOnOrAfter
```
and you want to get handle to it while processing `@node-saml/node-saml`'s return value of `validatePostResponseAsync`?
https://github.com/node-saml/node-saml/blob/e85389560a36b624ad5d399ee85e1c8a3b8adbea/src/saml.ts#L667-L669

If so ... return value's type is
https://github.com/node-saml/node-saml/blob/e85389560a36b624ad5d399ee85e1c8a3b8adbea/src/types.ts#L236-L251

and it is filled among other things with these information while processing validly signed auth response:
https://github.com/node-saml/node-saml/blob/e85389560a36b624ad5d399ee85e1c8a3b8adbea/src/saml.ts#L1163-L1165

I.e. IMHO this:
>...are there a way to get that back out from node-saml?...
>...but this seems a useful thing to be able to get in the validate callback or similar.

is already available for you via `getAssertionXml()` or  `getAssertion()`. NOTE: if assertion was encrypted this would contain encrypted content `getSamlResponseXml()`
